### DISCUSSION TITLE
Ensure F# Support

### DISCUSSION THREAD

**Discussion by TheCodeTraveler** [CommunityToolkit/Maui#235] [New Feature Discussions]
Let's ensure that `CommunityToolkit.Maui` has 1st Class support F# ❤️ 

In the Xamarin Community Toolkit, we learned that we must make a concerted effort to ensure that our library works with both F# and C# projects:
- https://github.com/xamarin/XamarinCommunityToolkit/issues/1124
- https://github.com/xamarin/XamarinCommunityToolkit/pull/1181
- https://github.com/xamarin/XamarinCommunityToolkit/pull/1133

I recommend making the following additions:
- Create `CommunityToolkit.Maui.Samples.Fsharp`
  - This ensures that the `CommunityToolkit.Maui` DLL compiles in F# projects
- Keep `CommunityToolkit.Maui.UnitTests`
  - This will remain our main Unit Test project
- Create `CommunityToolkit.Maui.UnitTests.FSharp`
  - This unit test project will exclusively test generated code for F# (because `CommunityToolkit.Maui.UnitTests` cannot accomplish that)

The new sample app, `CommunityToolkit.Maui.Samples.Fsharp`, doesn't need to replicate `CommunityToolkit.Maui.Sample`. Rather, it just needs to include a project reference to `CommunityToolit.Maui` to ensure that both `CommunityToolkit.Maui` and `CommunityToolkit.Maui.Core` successfully compile in an F# project. 

And the new Unit Test project, `CommunityToolkit.Maui.UnitTests.FSharp` doesn't need to replicate code from `CommunityToolkit.Maui.UnitTests`, rather it just needs to test code generated by Source Generators for F#.

**Comment by pictos** [comment]
If I'm not wrong just C# support Source Generators (SG), so we need to pay attention when adding SG features like
- #224 
and think how we will work around them

**Reply by TheCodeTraveler** [reply]
Yes, good catch! 

We should probably add an F# Unit Test Project too. 

I think this'll structure will allow us to continue to writing our UnitTests in C# while ensuring all Tests pass for F#:

1. Convert `CommunityToolkit.Maui.UnitTests` to a Shared Project
    - I.e. `CommunityToolkit.Maui.UnitTests.csproj` -> `CommunityToolkit.Maui.UnitTests.shproj`
2. Create `CommunityToolkit.Maui.UnitTests.Fsharp`
    - Include ProjectReference to `CommunityToolkit.Maui.UnitTests` Shared Project
3. Create `CommunityToolkit.Maui.UnitTests.Csharp`
    - Include ProjectReference to `CommunityToolkit.Maui.UnitTests` Shared Project

**Reply by TheCodeTraveler** [reply]
Dang - I just tried the above architecture in the [`Add-F#-Unit-Test-Project` branch](https://github.com/CommunityToolkit/Maui/tree/Add-F%23-Unit-Test-Project) and it doesn't work

tl;dr F# Projects don't compile `.cs` files

We can see in this screenshot that only UnitTests from the `CommunityToolkit.Maui.UnitTests.Csharp` project appear in the Test Explorer window.

![image](https://user-images.githubusercontent.com/13558917/148463906-49513d04-1300-4fea-8cd2-89d09cbd7590.png)

And when running `dotnet test` against `CommunityToolkit.Maui.UnitTests.Fsharp`, we get the following error:
> A total of 1 test files matched the specified pattern.
No test is available in /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.UnitTests.Fsharp/bin/Debug/net6.0/CommunityToolkit.Maui.UnitTests.Fsharp.dll.

```bash
> bramin@Brandons-MacBook-Pro CommunityToolkit.Maui.UnitTests.Fsharp % dotnet test
  Determining projects to restore...
  Restored /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.UnitTests.Fsharp/CommunityToolkit.Maui.UnitTests.Fsharp.fsproj (in 537 ms).
  2 of 3 projects are up-to-date for restore.
  CommunityToolkit.Maui.Core -> /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.Core/bin/Debug/net6.0/CommunityToolkit.Maui.Core.dll
  CommunityToolkit.Maui -> /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui/bin/Debug/net6.0/CommunityToolkit.Maui.dll
  CommunityToolkit.Maui.UnitTests.Fsharp -> /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.UnitTests.Fsharp/bin/Debug/net6.0/CommunityToolkit.Maui.UnitTests.Fsharp.dll
Test run for /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.UnitTests.Fsharp/bin/Debug/net6.0/CommunityToolkit.Maui.UnitTests.Fsharp.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
No test is available in /Users/bramin/GitHub/CommunityToolkit.Maui/src/CommunityToolkit.Maui.UnitTests.Fsharp/bin/Debug/net6.0/CommunityToolkit.Maui.UnitTests.Fsharp.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
```

**Reply by TheCodeTraveler** [reply]
Here's a new architecture that would still ensure we test all of our code for F# compatibility:
- Create `CommunityToolkit.Maui.Samples.Fsharp`
  - This ensures that the `CommunityToolkit.Maui` DLL compiles in F# projects
- Keep `CommunityToolkit.Maui.UnitTests`
  - This will remain our main Unit Test project
- Create `CommunityToolkit.Maui.UnitTests.FSharp`
  - This unit test project will exclusively test generated code for F# (because `CommunityToolkit.Maui.UnitTests` cannot accomplish that)

Going forward, all new PRs that include Source Generators will be required to add Unit Tests to both `CommunityToolkit.Maui.UnitTests` and `CommunityToolkit.Maui.UnitTests.FSharp`. And all PRs that don't include Source Generators would only need to add UnitTests to to `CommunityToolkit.Maui.UnitTests`.

**Reply by VladislavAntonyuk** [reply]
Looks like F# doesn't support source generators: https://github.com/fsharp/fslang-suggestions/issues/864

We depends on it, right?

**Comment by CliffAgius** [comment]
Closed as answered
